added new routes get and post at bubble-id/join, get shows the join p…

dmahfouda opened the pull request:
…age, posting to the route should allow new user to join

@jackrr Should I should try to reuse some of our existing login flow? In this case, I'll likely have to add arguments to your existing functions — primarily a bubble id, to indicate that a new member is intending to join a bubble in addition to being created.

Any thoughts or caveats on how to do the above?

Or should I create new function that handles a member who is logging in for the purpose of joining a specific bubble?

dmahfouda committed:
added new routes get and post at bubble-id/join

get shows the join page, posting to the route should allow new user to join

jackrr commented on src/bubble/bubbles.clj:
```diff
@@ -40,12 +41,16 @@
     (println (bubble-members param-id))
     (views/base-view
      [[:h1 (:bubbles/name bubble)]
+      [:h2 (str "Join bubble here: localhost:3000/bubble/" param-id "/join")]
```
Might be nice to include a link to a pre-filled SMS message. The following should work:

```clojure
(let [join-link (str "http://localhost:3000/bubble/" param-id "/join")]
  [:a {:href (str "sms:?&=" (ring.util.codec/form-encode {:body (str "Join my bubble: " join-link}))} (str "Bubble join link: " join-link)]
```

jackrr commented:
@dmahfouda I think the "right" way to go about it would be to change the app in such a way that the login code doesn't need to know about the "join" use case at all.

The way I'd build it is as follows:

1. Add a new /bubble/:bubbleId/join route that adds the currently logged-in user (via `(:current-user req)`) as a member to the bubble specified in the route param
2. Make the new handler from the above step redirect logged-out sessions to login (see [login/logged-in usage](https://github.com/jackrr/bubble/blob/dev/src/bubble/handler.clj#L13) for an example of a route that does this)
3. Improve the login flow so that it redirects back to the original route after successful login (this will lead to the user being added to the bubble via the default logic built in step 1)
  - This can be built by carrying a long a `redirect` query param through each step of the login flow

This way we can keep the login logic concerned only with login -- it doesn't need to worry about the manifold reasons _why_ a user might need to login.

Does the above make sense? Thoughts?

I would be happy to pair on implementing this, though I'm confident you know enough to do steps 1 and 2 comfortably and at least get started on 3.

jackrr commented on src/bubble/bubbles.clj:
```diff
@@ -40,12 +41,16 @@
+            (bubble-members param-id))]
+      [:form {:action param-id :method "post"}
+       [:input {:placeholder "Your phone" :name "userphone"}]
+       [:button () "join bubble"]]])))
```
I don't think this form belongs on this page, given that you already have to be logged in and a member of the bubble to view it.

dmahfouda replied:
Removed this!

jackrr commented on src/bubble/login.clj:
```diff
@@ -24,6 +24,21 @@
+                       [:input {:name "phone" :placeholder "Phone #"}]
+                       [:input {:name "short-code" :type "checkbox"}]
+                       [:label {:for "short-code"} "Send a code instead"]
+                       [:button {:name "submit"} "Send me a link"]]])))
```
If you agree with the flow I propose in this comment: https://github.com/jackrr/bubble/pull/3#issuecomment-846421253, you might want to change this page to be more of a "splash page" for the bubble that includes a button/link to join (also show bubble name, some "you have been invited blah blah blah" text). Clicking the button/link will navigate to `POST /bubbles/:id/join` -- if you're already logged in, you get added. If not, you get redirected to the existing login flow.

dmahfouda committed:
added count and name to bubble splash page

added ability to add users to bubble

jackrr commented on src/bubble/login.clj:
```diff
@@ -24,6 +24,10 @@
 
+
+(defn join-bubble [req]
+  )
```
Dead fn

dmahfouda replied:
Removed this—

dmahfouda committed:
readme updates

dmahfouda committed:
removed unneccessary form and dead function

dmahfouda committed:
whitespace

dmahfouda committed:
updated README

jackrr committed:
logs: remove prints and formatting

removes a few debugging print statements and fixes some unintentional
whitespace changes.
